keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1514 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This enhancement adds mutating admission policies, declared using CEL
+expressions, as an alternative to mutating admission webhooks. This continues
+the work started by [`KEP-3488`
+API](/keps/sig-api-machinery/3488-cel-admission-control/README.md) for
+validating admission policies.
+
+This enhancement proposes an approach where mutations are declared in CEL by
+combining CEL's object instantiation and Server Side Apply's merge algorithms.
+
+## Motivation
+
+A large proportion of mutating admission needs are for relatively simple
+operations such as setting a label, setting a field, or adding a sidecar
+container to a pod. These mutations can be expressed trivially in only a few
+lines of CEL, eliminating the developmental and operational complexity of a
+webhook.
+
+Offering CEL based mutation also has other fundamental advantages over webhooks.
+CEL mutations can be declared in a way that allows the kube-apiserver to
+introspect the mutation and extract useful information about which fields the
+mutation operation reads and writes. This information can be leveraged to do
+things like finding order for mutating admission policies that minimizes the
+need for reinvocation. Also, in-process mutation is sufficiently fast,
+especially when compared with webhooks, that it is reasonable to re-run
+mutations to do things like validate that multiple muation policy is still
+applied after all other mutating admission operations have been applied.
+
+### Goals
+
+- Provide an alternative to mutating webhooks for the vast majority of mutating
+  admission use cases.
+- Provide the in-tree extensions needed to build policy frameworks for
+  Kubernetes, again without requiring webhooks for the vast majority of use
+  cases.
+- Provide an out-of-tree implementation of this enhancement (using a webhook)
+  that is supported by the Kubernetes org to provide this enhancement
+  functionality to Kubernetes versions where this enhancement is not available.
+- Provide core functionality as a library so that use cases like GitOps,
+  CI/CD pipelines, and auditing can run the same CEL validation checks
+  that the API server does.
```

Comment:
Would this library likely also mean that unit testing could be performed on the policies away from a cluster? I think it would be advantageous to be able to demonstrate a method for testing the policies to potential users of the feature

Reply:
Thanks to the awesome contributors from undistro who put together the CEL playground: https://playcel.undistro.io/. We could potentially use this for checking the patch generated from CEL looks good or checking if the final stat meets the user's expectation. We would have similar tools set up separately as well.